src/models/geograph.py

```diff
@@ -518,22 +520,24 @@ def add_habitat(
                 invalid_indexes.append(node)
                 continue
 
-            polygon = polygons[node]
+            polygon = prep(polygons[node])
             if max_travel_distance > 0:
-                buff_poly = buff_polygons[node]
+                buff_poly_bounds = buff_polygons[node].bounds
+                buff_poly = prep(buff_polygons[node])
             else:
                 buff_poly = polygon
+                buff_poly_bounds = polygons[node].bounds
             # Query rtree for all polygons within `max_travel_distance` of the original
-            for nbr in self.rtree.intersection(buff_poly.bounds):
+            for nbr in self.rtree.intersection(buff_poly_bounds):
                 # If a node is not a habitat class node, don't add the edge
                 if hgraph.nodes[nbr]["class_label"] not in valid_classes_set:
```

Comment:
Could we drop the invalid classes as a bulk operation in hgraph at the outset of the function? That might give us another slight speed up (eg from avoiding the check and python-looping over unnecessary geoms here)

Reply:
Good point - I've added this and it makes the tqdm iteration counter more accurate but makes absolutely no difference to speed as far as I can tell. Turns out those checks are really fast and you have to do some upfront computation to get the indices of invalid nodes.